### PR TITLE
faster descriptions

### PR DESCRIPTION
**Pull request by andre-merzky**
This PR updates the RP description classes to faster and type-safe versions as discussed [here](https://github.com/radical-cybertools/radical.pilot/tree/prelim/descriptions/docs/architecture/dicts),

### CODE REVIEW

**Comment by andre-merzky** [comment]
@iparask , @lee212 : ping for review

**Comment by andre-merzky** [comment]
@iparask , @lee212 : ping for review
@mturilli : or maybe you have the time?

File: docs/architecture/dicts/spec_config.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+
+
+import radical.utils as ru
+import spec_attribs  as a
+
+
+class CFG(ru.Config):
```
Comment (mturilli) [inline docs/architecture/dicts/spec_config.py]: I am not sure I like this naming. I see it is concise but even after reading the documentation just 5 minutes ago I cannot remember what it stands for.

File: docs/architecture/dicts/spec_rumunch.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+               a.CPU_THREADS     : int         ,
+               a.CPU_THREAD_TYPE : str         ,
+               a.GPU_PROCESSES   : int         ,
+               a.GPU_PROCESS_TYPE: str         ,
+               a.GPU_THREADS     : int         ,
+               a.GPU_THREAD_TYPE : str         ,
+               a.LFS_PER_PROCESS : int         ,
+               a.MEM_PER_PROCESS : int         ,
+    }
+
```
Comment (mturilli) [inline docs/architecture/dicts/spec_rumunch.py]: Neat! :)

File: docs/architecture/dicts/spec_schema.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+class SCH(dict):
+
+    _any    = Or(object, int, float, None)
+    _schema = Schema({
+                      a.EXECUTABLE         : str,
+             Optional(a.KERNEL          )  : str,
+             Optional(a.NAME            )  : str,
+             Optional(a.ARGUMENTS       )  : [str],
+             Optional(a.ENVIRONMENT     )  : Or({}, {str: str}),
```
Comment (mturilli) [inline docs/architecture/dicts/spec_schema.py]: Double neat :D

File: docs/architecture/dicts/t.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+#     list_of_ints : List[int]
+#
+#
+# m = Model(age=42, list_of_ints=[1, '2', b'3'])
+# m.age = True
+# # print(m.middle_name)  # not a model field!
+# Model()
+
+
+from good.voluptuous import *
```
Comment (mturilli) [inline docs/architecture/dicts/t.py]: This cracked me up

File: docs/source/overview.rst [hunk 5]
```diff
@@ -112,14 +112,14 @@ circles). Application contains pilot and CU descriptions; RP Client has two
 components: Pilot Manager and Unit Manager. Pilot descriptions are passed to
 the Pilot Manager and Unit descriptions to the Unit Manager. The Pilot Manager
 uses Pilot Launcher to launch 2 of the 5 described pilots. One pilot is
-submitted to the Local Resource Management System (LRMS) of Resource A, the
-other pilot to the LRMS of Resource B. Once instantiated, each pilot becomes
+submitted to the local Resource Management system (RM) of Resource A, the
```
Comment (mturilli) [inline docs/source/overview.rst]: `Resource Management system (RM)`->`Resource Management (RM) system`

File: examples/00_getting_started.py [hunk 6]
```diff
@@ -57,25 +56,23 @@
                    'project'       : config[resource]['project'],
                    'queue'         : config[resource]['queue'],
                    'access_schema' : config[resource]['schema'],
-                   'cores'         : config[resource]['cores'],
+                   'cores'         : 128
```
Comment (mturilli) [inline examples/00_getting_started.py]: I do not think we should hard-code this

File: examples/00_getting_started.py [hunk 7]
```diff
@@ -57,25 +56,23 @@
...
                   }
         pdesc = rp.ComputePilotDescription(pd_init)
 
         # Launch the pilot.
         pilot = pmgr.submit_pilots(pdesc)
 
-        report.header('submit units')
+        n = 1024  # number of units to run
```
Comment (mturilli) [inline examples/00_getting_started.py]: the same here

File: src/radical/pilot/agent/lm/__init__.py [hunk 8]
```diff
@@ -3,5 +3,5 @@
 __license__   = "MIT"
 
 from .base import *
-from .base import LaunchMethod 
+from .base import LM
```
Comment (mturilli) [inline src/radical/pilot/agent/lm/__init__.py]: Unsurprisingly, I liked the previous extended naming much better. What will be `LM` for a newbie? What will be for me in a week? As we say in Italian `Boh!`...

File: src/radical/pilot/agent/rm/__init__.py [hunk 9]
```diff
@@ -4,5 +4,5 @@
 
 
 from .base import *
-from .base import LRMS as RM
+from .base import RM
```
Comment (mturilli) [inline src/radical/pilot/agent/rm/__init__.py]: Comment about LM applies also to RM.

File: examples/00_getting_started.py (hunk 7 above)
Comment (andre-merzky) [reply examples/00_getting_started.py]: This was always hard coded, and I'd like to leave that for now, to keep the examples simple (only resource details are in the config right now).

File: src/radical/pilot/agent/lm/__init__.py (hunk 8 above)
Comment (andre-merzky) [reply src/radical/pilot/agent/lm/__init__.py]: I reviewed this again: the above was a consistency fix I stumbled over when renaming LRMS to RM.  Consistently renaming LM to LaunchMethod and RM to ResourceManager touches 55 files in about 250 places, and then some documentation.  I would thus like to leave that out of this PR and do that (with lower priority) in a separate refactoring.  If that works for you, would you mind opening a ticket?

File: examples/00_getting_started.py (hunk 7 above)
Comment (mturilli) [reply examples/00_getting_started.py]: OK, works for me.

**Comment by mturilli** [comment]
All the Travis tests fails. Could we fix that before merging?

File: src/radical/pilot/agent/lm/__init__.py (hunk 8 above)
Comment (andre-merzky) [reply src/radical/pilot/agent/lm/__init__.py]: The LM/RM refactoring is done now.

**Comment by andre-merzky** [comment]
The LM -> LaunchMethod / RM -> ResourceManager renaming has been done now, can you have a look?  Thanks!

**Pull request merged**